File: app/src/test/java/com/fractalwrench/crazycats/image/PresenterLifecycleTest.java
package com.fractalwrench.crazycats.image;

import android.support.annotation.NonNull;

import org.junit.Before;
import org.junit.Test;

import java.io.IOException;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertNull;

/**
 * Ensures that a {@link Presenter} responds to lifecycle methods as expected.
 * <p>
 * More precisely, there should never be imbalanced calls to {@link Presenter#start(V)}
 * and {@link Presenter#stop()}, and the View should always be dereferenced after
 * presentation ends
 *
 * @param <V> the view type
 * @param <P> the presenter type
 */
public abstract class PresenterLifecycleTest<V, P extends Presenter<V>> {

    private P presenter;

    @Before
    public void setup() throws IOException {
        presenter = getContentViewPresenter();
        assertNotNull(presenter);
    }

    /**
     * Supply the Presenter to be tested
     */
    @NonNull
    protected abstract P getContentViewPresenter() throws IOException;

    @Test(expected = IllegalArgumentException.class)
    public void testNullView() {
        //noinspection ConstantConditions
        presenter.start(null);
    }

    @Test
    public void testStartStop() {
        assertNull(presenter.contentView);
        presenter.start(getContentView());
        assertEquals(presenter.contentView, presenter.contentView);

        presenter.stop();
        assertNull(presenter.contentView);
    }

    /**
     * Supplies the Content View to be tested.
     * <p>
     * Should be a mocked interface that can run on a JVM without instrumentation.
     */
    @NonNull
    protected abstract V getContentView();

    @Test(expected = IllegalStateException.class)
    public void testStopFirst() {
        presenter.stop();
    }

    @Test(expected = IllegalStateException.class)
    public void testStartTwice() {
        presenter.start(getContentView());
        presenter.start(getContentView());
    }

    @Test(expected = IllegalStateException.class)
    public void testStopTwice() {
        presenter.start(getContentView());
        presenter.stop();
        presenter.stop();
    }

}

File: app/src/test/java/com/fractalwrench/crazycats/image/PresenterLifecycleTest.kt
package com.fractalwrench.crazycats.image

import junit.framework.Assert.*
import org.junit.Before
import org.junit.Test
import java.io.IOException

/**
 * Ensures that a [Presenter] responds to lifecycle methods as expected.
 *
 *
 * More precisely, there should never be imbalanced calls to [Presenter.start]
 * and [Presenter.stop], and the View should always be dereferenced after
 * presentation ends

 * @param <V> the view type
 * *
 * @param </V><P> the presenter type
</P> */
abstract class PresenterLifecycleTest<V, out P : Presenter<V>> {

    private var presenter: P? = null

    @Before
    @Throws(IOException::class)
    fun setup() {
        presenter = getContentViewPresenter()
        assertNotNull(presenter)
    }

    /**
     * Supply the Presenter to be tested
     */
    protected abstract fun getContentViewPresenter(): P

    @Test
    fun testStartStop() {
        assertNull(presenter!!.contentView)
        presenter!!.start(contentView)
        assertEquals(presenter!!.contentView, presenter!!.contentView)

        presenter!!.stop()
        assertNull(presenter!!.contentView)
    }

    /**
     * Supplies the Content View to be tested.
     *
     *
     * Should be a mocked interface that can run on a JVM without instrumentation.
     */
    protected abstract val contentView: V

    @Test(expected = IllegalStateException::class)
    fun testStopFirst() {
        presenter!!.stop()
    }

    @Test(expected = IllegalStateException::class)
    fun testStartTwice() {
        presenter!!.start(contentView)
        presenter!!.start(contentView)
    }

    @Test(expected = IllegalStateException::class)
    fun testStopTwice() {
        presenter!!.start(contentView)
        presenter!!.stop()
        presenter!!.stop()
    }

}